index.js

```diff
@@ -92,8 +92,32 @@ function serveStatic (root, options) {
       path = ''
     }
 
+    // create send options
+    var sendOpts = Object.create(opts)
+    var _root
+    if (typeof root === 'function') {
+      try {
+        _root = root(req)
+      } catch (e) {
+        next(e)
+        return
+      }
+    } else {
+      _root = root
+    }
+
+    if (!_root) {
+      // ignore fallthrough option
+      // as this should be considered a major error
+      var error = createError(404)
```

Comment:
imo this should be a "major error", like your comment. This means it should return a 500 and probably have a useful message so the developer can understand their `root` function did not return a value. This will basically 404 silently.